rust_snuba/src/strategies/python.rs

```diff
@@ -81,82 +81,84 @@ impl PythonTransformStep {
-                } => (original_message_meta, result),
-            };
-            match message_result {
-                Ok(data) => {
-                    let replacement = BytesInsertBatch {
-                        rows: data.rows,
-                        // TODO: Actually implement this
-                        commit_log_offsets: BTreeMap::from([]),
-                    };
-
-                    if let Err(SubmitError::MessageRejected(MessageRejected {
-                        message: transformed_message,
-                    })) = self
-                        .next_step
-                        .submit(original_message_meta.replace(replacement))
-                    {
-                        self.message_carried_over = Some(transformed_message);
-                    }
-                }
-                Err(error) => {
-                    tracing::error!(error, "Invalid message");
+        if let Some(message_carried_over) = self.message_carried_over.take() {
+            if let Err(SubmitError::MessageRejected(MessageRejected { message })) =
+                self.next_step.submit(message_carried_over)
+            {
+                self.message_carried_over = Some(message);
+                return;
+            }
+        }
+
+        debug_assert!(self.message_carried_over.is_none());
+
+        let (original_message_meta, message_result) = match self.handles.pop_front() {
+            Some(TaskHandle::Procspawn {
+                original_message_meta,
+                join_handle,
+            }) => {
+                let handle = join_handle.into_inner().unwrap();
+                let result = handle.join().expect("procspawn failed");
+                (original_message_meta, result)
+            }
+            Some(TaskHandle::Immediate {
+                original_message_meta,
+                result,
+            }) => (original_message_meta, result),
+            None => return,
+        };
+
+        match message_result {
+            Ok(data) => {
+                let replacement = BytesInsertBatch {
+                    rows: data.rows,
+                    // TODO: Actually implement this
+                    commit_log_offsets: BTreeMap::from([]),
+                };
+
+                if let Err(SubmitError::MessageRejected(MessageRejected {
+                    message: transformed_message,
+                })) = self
+                    .next_step
+                    .submit(original_message_meta.replace(replacement))
+                {
+                    self.message_carried_over = Some(transformed_message);
                 }
             }
+            Err(error) => {
+                tracing::error!(error, "Invalid message");
+            }
         }
     }
 }
 
 impl ProcessingStrategy<KafkaPayload> for PythonTransformStep {
     fn poll(&mut self) -> Result<Option<CommitRequest>, InvalidMessage> {
-        self.check_for_results();
+        if self.handles.len() > self.max_queue_depth {
```

Comment:
Doesn't this mean we are *never* checking for results if we never hit max queue depth? How would that work for low volume topics/environments?

Reply:
ah you're right. i was relying on `join()` to eventually flush things out, but this doesn't work. how about a parameter called `max_queue_latency`?

Reply:
I think that approach sounds good. Wdyt about not even making it a parameter and just hardcoding 1 second or something? I worry that having too many parameters for stuff makes it confusing/difficult for users to pick good values and it seems that a constant value here would be just fine at any volume really.

Reply:
done. i set it to 10 seconds because otherwise it has a lot of impact on throughput